tests/sentry/backup/__init__.py

```diff
@@ -12,114 +13,129 @@
+            # referenced by foreign key. Those do not explicitly need to be set - we don't care that
+            # models that reference this model exist, just that this model exists in its most
+            # filled-out form.
+            if isinstance(f, models.ManyToOneRel) or isinstance(f, models.ManyToManyRel):
+                continue
+
+            # TODO(getsentry/team-ospo#156): For some reason we currently don't always serialize
+            # some `ManyToManyField`s with the `through` property set. I'll investigate, but will
+            # skip over these for now.
+            if isinstance(f, models.ManyToManyField):
+                continue
+
+            if not isinstance(f, models.Field):
+                continue
+            if field_name not in data:
+                mistakes.append(f"Must include field: `{field_name}`")
+                continue
+            if f.has_default():
+                default_value = f.get_default()
+                serialized = encoder.encode(default_value)
+                if serialized == data:
+                    mistakes.append(f"Must use non-default data: `{field_name}`")
+                    return
+
+        # If one model instance has N mistakes, and another has N - 1 mistakes, we want to keep the
+        # shortest list, to give the user the smallest number of fixes to make when reporting the
+        # mistake.
+        if name not in mistakes_by_model or (len(mistakes) < len(mistakes_by_model[name])):
+            mistakes_by_model[name] = mistakes
+
+    for name, mistakes in mistakes_by_model.items():
+        num = len(mistakes)
+        if num > 0:
+            raise AssertionError(f"Model {name} has {num} mistakes: {mistakes}")
+
+
+def expect_models(group: set[NormalizedModelName], *marking: Type | Literal["__all__"]) -> Callable:
     """
-    A function that runs at module load time and marks all models that appear in a given test function.
+    A function that runs at module load time and marks all models that appear in a given test
+    function. The first argument stores the tracked models in a global group, which we then check in
+    `test_coverage.py` for completeness.
 
     Use the sentinel string "__all__" to indicate that all models are expected.
     """
 
     all: Literal["__all__"] = "__all__"
+    target_models: Set[Type[models.Model]] = set()
     for model in marking:
         if model == all:
             all_models = get_exportable_sentry_models()
             group.update({get_model_name(c) for c in all_models})
-            return list(all_models)
+            target_models = all_models.copy()
+            break
 
         group.add(get_model_name(model))
-    return marking
+        target_models.add(model)
+
+    def decorator(func: Callable):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            return func(*args, target_models, **kwargs)
```

Comment:
I'd also discourage signature-modifying decorators -- they're still action-at-a-distance and it'd be much more explicit to use a context manager

Reply:
The reason I want to use a decorator here is because `test_coverage.py` checks these for completeness statically. If we use a context manager, the `..._TESTED` globals only get populated at test runtime.

Reply:
is there another way you can satisfy that without metaprogramming?

Reply:
I suppose there always is, but it's a bigger lift than this change as constituted.

The main case we're trying to guard against is developers not thinking about backups when they add/modify models, thereby causing new models to be untested. If we tested all of the cases in a single large test then meta-programming would be unnecessary and we could just check at test runtime, but that would make for very large and unwieldy test cases (ex: we have ~10 collision tests that use `COLLISION_TESTED`, which make much more sense when split by model than if they were glued together into a single giga-test).

Reply:
Hi, just following up here. While the change as constituted may not be perfect, it is better than the current state of affairs (simpler decorator functionality), and fixes two relatively more pressing issues:

1. Unblocking future python version bumps by not returning from tests.
2. Providing devs with a clearer error message when they trip this check in CI.